### ISSUE TITLE
Use IReadonlyDictionary<TKey, TValue> instead of IEnumerable<KeyValuePair<TKey, TValue>>

### ISSUE THREAD

**Issue by gideonkorir** [opentracing/opentracing-csharp#44]
I'm proposing that we change methods that use ```IEnumerable<KeyValuePair<TKey,TValue>>``` to ```IReadonlyDictionary<TKey,TValue>```. I believe that using ```IReadonlyDictionary``` maintains the semantics  of an immutable map but is easier to work with e.g. we can use indexer ```map[key]``` and also access ```Count``` property.

If agreed we can make the following changes:

```
interface ISpanContext {
 IReadonlyDictionary<string, string> BaggageItems { get; }
}

interface ISpan
{
  Log(IReadonlyDictionary<string, string> fields);
  Log(DateTimeOffset timestamp, IReadonlyDictinary<string, string> fields)
}

````

Using ```IReadonlyDictionary<TKey,TValue>``` simplifies implementations e.g. 

1. Duplicate key behavior is undefined for ```Log(IEnumerable)``` so using ```IReadOnlyDictionary``` averts that behavior
2. In the ```Log(IEnumerable)``` it's simpler to check for semantic fields e.g. Brave library checks for ```event``` key

**Comment by gideonkorir** [comment]
Maybe we also change ```string ISpan.GetBaggageItem(string key)``` to ```bool ISpan.TryGetBaggageItem(string key, out string value)```. This makes explicit that the key could be missing and doesn't require the client to check for nulls.

**Comment by ndrwrbgs** [comment]
As duplicate keys are by definition undefined, it seems to make sense to make the contract represent that fact (ICollection<KeyValuePair> is slightly semantically different from IReadOnlyDictionary<KeyValuePair>).

[Note c#’s IReadOnlyDictionary isn’t actually enforced to be immutable, for those coming from Java it may be more clear to think of it as IReadableDictionary — and interface shouldn’t say what an implementor CAN’T do, but that’s for another time].

This change would be more in line with what the java implementation does, which is expose a map).

The one con of this change is that it adds an implementation constraint that is not present in the specification. The specification does bring forth the question of duplicate keys like this would address, but it does not require that baggage items, on transfer, be materialized (e.g. imagine passing around the baggage serialized, and being able to streaming-wise deserialize it. Based PURELY on the specification, an implementor would be capable of making every call to BaggageItems make a database call, but if we expose IReadOnlyDictionary a caller would expect that to NOT be the case - [index] calls should be quick).

Given that the specification does expose retrieving BaggaeItems by Key on a Span though, I think this would be a net positive change.

**Comment by ndrwrbgs** [comment]
> Maybe we also change string ISpan.GetBaggageItem(string key) to bool ISpan.TryGetBaggageItem(string key, out string value). This makes explicit that the key could be missing and doesn't require the client to check for nulls

Null should probably not be used to indicate no value, as null itself is a value (it's a value that represents no value, but there's a slight semantic difference in a key missing altogether, and a key being present with it's value being missing).

I swore I just read something that was discussing this yesterday on the specification/java pages, but searches today are not finding it. Perhaps the internet ate it 😄.

For a quick example of where this could be useful, take the following pseudo-code
```
var value = spanContext.GetBaggage(key);
if (value == null)
{
  var newValue = someComplicatedExpensiveIshMethod();
  // let's say newValue is null here
  spanContext.SetBaggage(key, newValue);
}
```
If this is called frequently, the developer would expect that the inner code is only being called once, but given today's API it would be called every time.

**Comment by gideonkorir** [comment]
@ndrwrbgs whilst the specification doesn't explicitly tell us how to handle duplicate keys, I think that the ```ISpan.GetBaggageItem``` & ```ISpan.SetTag``` implicitly expect single keys in the baggage items otherwise the 1st API should return ```IEnumerable<string>``` and the 2nd API should be ```ISpan.AddTag```.

The idea about the un-serialized baggage item is interesting but one could make an implementation of ```IReadOnlyDictionary``` that deserializes on access to any of it's APIs. Also we expect the client to have as few baggage items as possible; I think the simplicity of using ```IReadOnlyDictionary``` far outweighs the deserialization costs.

Your comment about ```null``` is spot on :), if we agreed to change baggage items to ```IReadOnlyDictionary``` I'd probably change ```ISpanContext``` to:

```
public interface ISpanContext
{
  IReadOnlyDictionary<string,string> BaggageItems { get; }
}

```
The readonly property has the same semantics as the currently defined method.

**Comment by cwe1ss** [comment]
Sorry for not replying - I’m on a sabbatical right now. Please see https://github.com/opentracing/opentracing-csharp/issues/52 for a topic that IMO should be resolved before we do any further changes.

**Comment by cwe1ss** [comment]
There's two sides to this topic:

#### 1. What type should be used for passing key:value pairs INTO the API. 
This currently affects the `Log`-method which accepts `IEnumerable<KeyValuePair<string, object>>`. 
* __Advantage:__ "Instrumentation"-code can pass a List/array instead of a dictionary which is faster as they don't do uniqueness checks on add. This is great IMO as these log-objects are created manually most of the time anyway so there's no reason to have duplicate keys. People can however still pass dictionaries if they want.
* __Disadvantage:__ This API doesn't enforce unique keys. The API therefore must define the behavior of duplicate keys as "undefined" and every "Implementation"-code has to deal with it somehow. That's not complicated but it can lead to different behavior in different implementations. 

#### 2. What type should be used to get key:value pairs FROM the API

This currently only affects `ISpanContext.GetBaggageItems` which also returns  `IEnumerable<KeyValuePair<string, object>>`.

* __Advantage:__ "Implementation"-code can implement storage however they want. They e.g. can use linked lists and 'yield return'. That's how .NET's [Activity type does it](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L114). With I*Dictionary an implementation would either have to use dictionaries internally as well or copy all the data into a dictionary in this method.
* __Disadvantages:__ 
  * It doesn't enforce uniqueness so implementations could return duplicate keys in theory. 
  * It doesn't allow using indexers. This could also be solved by adding a separate `GetBaggageItem(string key)` method, which already exists on `ISpan`.

So for me it's about performance/flexibility vs. strictness/usability. Do we want to give instrumentations and implementations more power or do we want to have a more "controlled" environment by enforcing a more strict contract. 

Obviously, should we have to change this again later for whatever reason, it would be easier to move from IDictionary to IEnumerable (as this is a base class) than from IEnumerable to IDictionary. 

------------

> In the Log(IEnumerable) it's simpler to check for semantic fields e.g. Brave library checks for event key

It's simpler but less efficient. Implementations have to go through every key:value pair anyway so checking for multiple keys by always doing a lookup is less efficient than iterating through everything just once and comparing keys.

> Maybe we also change string ISpan.GetBaggageItem(string key) to bool ISpan.TryGetBaggageItem(string key, out string value). This makes explicit that the key could be missing and doesn't require the client to check for nulls.

@gideonkorir, good idea. Could you please propose this in a separate issue? I want to make sure we concentrate on one topic here.

**Comment by cwe1ss** [comment]
#50 (which would ensure parity with Java as much as possible) would change `Log` to use `IDictionary` and would keep `IEnumerable` for `GetBaggageItems`. I guess that's the best short-term approach because being on par with Java is more important for now. We can discuss a change in separate issues together with the java people and it would be easy to move back to `IEnumerable`  should we decide to do so.

Another thing: Even if `Log` uses `IDictionary`, implementation code still has to check for duplicates internally as one could call `Log` multiple times.

**Comment by carlosalberto** [comment]
> #50 (which would ensure parity with Java as much as possible) would change Log to use IDictionary and would keep IEnumerable for GetBaggageItems. I guess that's the best short-term approach because being on par with Java is more important for now.

Agreed on that. I *think* it's a good idea to have this parity in place first, and then iterate over the items/issues as we go.

**Comment by ndrwrbgs** [comment]
Just a note, using `IDictionary` doesn't actually prevent duplicates either. The interface can't mandate no duplicates, it represents a collection that _logically_ does not have duplicates, but `Dictionary` has a custom `IEqualityComparer` passed in that isn't exposed on the interface, which means if the consumer plans to _do_ anything with these items that requires them to not be duplicates they still have to do that work.

TL;DR anything you can do (including duplicates) with `IEnumerable<KeyValuePair>` you can also do with `IDictionary` the only difference being `IDictionary` provides you a way to ask for a specific key (how the actual implementation does this it not really the callers concern) which allows the implementation to optimize, if possible (e.g. if it has stored items by keys, if it caches lookups for keys, etc).

It's worth noting we're not prohibited from introducing our own data type. I'd not recommend it, but nothing stops it :)
```
interface IEnumerableWithLookup<TKey, TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
{
  TValue GetValue(TKey key);
}
```

P.S. For ease of discussion I'm using `IDictionary` to refer to either `IDictionary` or `IReadOnlyDictionary`, as the discussion seems to have moved from there and if we decide to use dictionaries at all THEN we will have that discussion I think.

**Comment by cwe1ss** [comment]
I'm going to close this for the following reasons:
* It discusses two separate things (Parameter passed to `Log()` AND return type for `GetBaggageItems()`)
* If we change the parameter type for `Log()` then we should move to the most general type, which is `IEnumerable<KeyValuePair<string, object>>`. I created #91 for this purpose. Please file any comments/concerns there.
* Returning `I(ReadOnly)Dictionary` on `GetBaggageItems` is problematic because:
  * There already is the existing `ISpan.GetBaggageItem(string key)` method for accessing baggage by key. #86 discusses whether there should be the same method on `ISpanContext`
  * Returning a dictionary type would lead to different behavior for the same semantic thing: The existing `ISpan.GetBaggageItem(string key)` method returns `null` if the key doesn't exist but the `I(ReadOnly)Dictionary` indexer would throw an exception in that case.
  * Feel free to create a new issue (with details on these concerns) if you still think this should be changed.
